pkg/components/master_caches.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/ytsaurus/yt-k8s-operator/pkg/apiproxy"
+	"github.com/ytsaurus/yt-k8s-operator/pkg/consts"
+	"github.com/ytsaurus/yt-k8s-operator/pkg/labeller"
+	"github.com/ytsaurus/yt-k8s-operator/pkg/resources"
+	"github.com/ytsaurus/yt-k8s-operator/pkg/ytconfig"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type masterCache struct {
+	componentBase
+	server  server
+	initJob *InitJob
+}
+
+func NewMasterCache(cfgen *ytconfig.Generator, ytsaurus *apiproxy.Ytsaurus) Component {
+	resource := ytsaurus.GetResource()
+	l := labeller.Labeller{
+		ObjectMeta:     &resource.ObjectMeta,
+		APIProxy:       ytsaurus.APIProxy(),
+		ComponentLabel: consts.YTComponentLabelMasterCache,
+		ComponentName:  "Master",
+		MonitoringPort: consts.MasterCachesMonitoringPort,
+		Annotations:    resource.Spec.ExtraPodAnnotations,
+	}
+
+	server := newServer(
+		&l,
+		ytsaurus,
+		&resource.Spec.MasterCaches.InstanceSpec,
+		"/usr/bin/ytserver-master-cache",
+		"ytserver-master-cache.yson",
+		cfgen.GetMasterCachesStatefulSetName(),
+		cfgen.GetMasterCachesServiceName(),
+		cfgen.GetMasterCachesConfig,
+	)
+
+	initJob := NewInitJob(
+		&l,
+		ytsaurus.APIProxy(),
+		ytsaurus,
+		resource.Spec.ImagePullSecrets,
+		"default",
+		consts.ClientConfigFileName,
+		resource.Spec.CoreImage,
+		cfgen.GetNativeClientConfig)
+
+	return &masterCache{
+		componentBase: componentBase{
+			labeller: &l,
+			ytsaurus: ytsaurus,
+			cfgen:    cfgen,
+		},
+		server:  server,
+		initJob: initJob,
+	}
+}
+
+func (mc *masterCache) IsUpdatable() bool {
+	return true
+}
+
+func (mc *masterCache) Fetch(ctx context.Context) error {
+	return resources.Fetch(ctx, mc.server)
+}
+
+func (mc *masterCache) doSync(ctx context.Context, dry bool) (ComponentStatus, error) {
+	var err error
+
+	if ytv1.IsReadyToUpdateClusterState(mc.ytsaurus.GetClusterState()) && mc.server.needUpdate() {
+		return SimpleStatus(SyncStatusNeedFullUpdate), err
+	}
+
+	if mc.ytsaurus.GetClusterState() == ytv1.ClusterStateUpdating {
+		if status, err := handleUpdatingClusterState(ctx, mc.ytsaurus, mc, &mc.componentBase, mc.server, dry); status != nil {
+			return *status, err
+		}
+	}
+
+	if mc.server.needSync() {
+		if !dry {
+			err = mc.doServerSync(ctx)
+		}
+		return WaitingStatus(SyncStatusPending, "components"), err
+	}
+
+	if !mc.server.arePodsReady(ctx) {
+		return WaitingStatus(SyncStatusBlocked, "pods"), err
+	}
+
+	return mc.initJob.Sync(ctx, dry)
+}
+
+func (mc *masterCache) Status(ctx context.Context) ComponentStatus {
+	status, err := mc.doSync(ctx, true)
+	if err != nil {
+		panic(err)
+	}
+
+	return status
+}
+
+func (mc *masterCache) Sync(ctx context.Context) error {
+	_, err := mc.doSync(ctx, false)
+	return err
+}
+
+func (mc *masterCache) doServerSync(ctx context.Context) error {
+	statefulSet := mc.server.buildStatefulSet()
+	mc.addAffinity(statefulSet)
+	return mc.server.Sync(ctx)
+}
+
+func (mc *masterCache) addAffinity(statefulSet *appsv1.StatefulSet) {
```

Comment:
let's create function for adding affinity to stateful set in helpers.go (a maybe create affinity.go) for example and use it in master and mastercache components.

Reply:
Done